controllers/progressiverollout_controller.go

```diff
@@ -71,67 +72,119 @@ func (r *ProgressiveRolloutReconciler) Reconcile(ctx context.Context, req ctrl.R
 		}
-		r.Log.V(1).Info("apps selected", "apps", fmt.Sprintf("%v", apps))
+		r.Log.V(1).Info("apps selected", "apps", utils.GetAppsName(apps))
 
 		// Remove the annotation from the OutOfSync Applications before passing them to the Scheduler
 		// This action allows the Scheduler to keep track at which stage an Application has been synced.
-		outOfSyncApps := utils.FilterAppsBySyncStatusCode(apps, argov1alpha1.SyncStatusCodeOutOfSync)
-		if err = r.removeAnnotationFromApps(outOfSyncApps, utils.ProgressiveRolloutSyncedAtStageKey); err != nil {
-			return ctrl.Result{}, err
+		outOfSyncApps := utils.GetAppsBySyncStatusCode(apps, argov1alpha1.SyncStatusCodeOutOfSync)
+		if rErr := r.removeAnnotationFromApps(outOfSyncApps, utils.ProgressiveRolloutSyncedAtStageKey); rErr != nil {
+			message := "unable to remove out-of-sync annotation"
+			log.Error(rErr, message)
+			if err := r.updateStageStatus(stage.Name, message, deploymentskyscannernetv1alpha1.PhaseUnknown, &pr); err != nil {
+				return ctrl.Result{}, err
+			}
+			return ctrl.Result{}, rErr
 		}
 
 		// Get the Applications to update
+		// TODO: update the scheduler to return the InProgress apps
 		scheduledApps := scheduler.Scheduler(apps, stage)
 
+		// TODO: status status - update targets
+
 		for _, s := range scheduledApps {
 			r.Log.Info("syncing app", "app", s)
 
-			_, err := r.syncApp(s.Name)
+			_, sErr := r.syncApp(s.Name)
+
+			if sErr != nil {
+				if !strings.Contains(sErr.Error(), "another operation is already in progress") {
+					message := "unable to sync app"
+					log.Error(sErr, message, "message", sErr.Error())
+					if err := r.updateStageStatus(stage.Name, message, deploymentskyscannernetv1alpha1.PhaseUnknown, &pr); err != nil {
+						return ctrl.Result{}, err
+					}
+
+					// TODO: stage status - update failed clusters
 
-			if err != nil {
-				if !strings.Contains(err.Error(), "another operation is already in progress") {
-					log.Error(err, "unable to sync app", "message", err.Error())
-					return ctrl.Result{}, err
+					return ctrl.Result{}, sErr
 				}
 			}
 		}
 
-		if scheduler.IsStageFailed(apps, stage) {
-			// TODO: updated status
-			r.Log.Info("stage failed")
+		if scheduler.IsStageFailed(apps) {
+			message := "stage failed"
+			r.Log.Info(message)
+			if err := r.updateStageStatus(stage.Name, message, deploymentskyscannernetv1alpha1.PhaseFailed, &pr); err != nil {
+				return ctrl.Result{}, err
+			}
+
+			failedMessage := fmt.Sprintf("%s stage failed", stage.Name)
+			failed := pr.NewStatusCondition(deploymentskyscannernetv1alpha1.CompletedCondition, metav1.ConditionTrue, deploymentskyscannernetv1alpha1.StagesFailedReason, failedMessage)
+			apimeta.SetStatusCondition(pr.GetStatusConditions(), failed)
+			if err := r.patchStatus(ctx, &pr); err != nil {
+				r.Log.Error(err, "failed to update object status")
+				return ctrl.Result{}, err
+			}
+			r.Log.Info("rollout failed")
+			// We can set Requeue: true once we have a timeout in place
```

Comment:
Currently, stage failed fails the rollout and doesn't retry - this doesn't leave much space for 'eventual consistency'. However, until we have a `progressDeadlineSeconds` in place, I'd keep it this way to avoid infinite loops.